props/websubmit/src/main.rs

```diff
@@ -177,6 +177,112 @@ fn run_sc_policy(ctx: Arc<Context>) -> Result<()> {
+        AuthDisclosureProp { cx }
+    }
+
+    pub fn check(self) {
+        for c_id in self.cx.desc().controllers.keys() {
+            // All srcs that have no influencers
+            let roots = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|n| {
+                    self.cx
+                        .influencers(*n, paralegal_policy::EdgeType::DataAndControl)
+                        .next()
+                        .is_none()
+                })
+                .collect::<Vec<_>>();
+
+            let safe_scopes = self
+                .cx
+                // All nodes marked "safe"
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|n| self.cx.has_marker(marker!(safe_source), *n))
+                // And all nodes marked "safe_with_bless"
+                .chain(self.cx.all_nodes_for_ctrl(*c_id).filter(|node| {
+                    self.cx.has_marker(marker!(safe_source_with_bless), *node)
+                        && self
+                            .cx
+                            // That are influenced by a node marked "bless"
+                            .influencers(*node, paralegal_policy::EdgeType::DataAndControl)
+                            .any(|b| self.cx.has_marker(marker!(bless_safe_source), b))
+                }))
+                .collect::<Vec<_>>();
+            let sinks = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|n| self.cx.has_marker(marker!(sink), *n))
+                .collect::<Vec<_>>();
+            let mut sensitives = self
+                .cx
+                .all_nodes_for_ctrl(*c_id)
+                .filter(|node| self.cx.has_marker(marker!(sensitive), *node));
+
+            let controller_valid = sensitives.all(|sens| {
+                sinks.iter().all(|sink| {
+                    // sensitive flows to store implies
+                    !(self
+                        .cx
+                        .flows_to(sens, *sink, paralegal_policy::EdgeType::Data))
+                        || sink.associated_call_site().is_some_and(|sink_callsite| {
+                            // scopes for the store
+                            let store_scopes = self
+                                .cx
+                                .influencers(sink_callsite, paralegal_policy::EdgeType::Data)
+                                .filter(|n| self.cx.has_marker(marker!(scopes), *n))
+                                .collect::<Vec<_>>();
+
+                            // all flows are safe before scope
+                            let safe_before_scope = self.cx.always_happens_before(
+                                roots.iter().cloned(),
+                                |n| safe_scopes.contains(&n),
+                                |n| store_scopes.contains(&n)).unwrap();
+
+                            let safe = !store_scopes.is_empty()
+                                && safe_before_scope.holds();
```

Comment:
You may consider calling `AlwaysHappensBefore::report()`, which will emit an error for you and also warn for potential vacuities.

And before that I would suggest an `assert_error!(self.cx, !store_scopes.is_empty(), "Meaningful error message about not finding any scopes")`

Reply:
done